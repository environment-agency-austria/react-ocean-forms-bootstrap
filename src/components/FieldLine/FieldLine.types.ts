import { IUseFieldResult } from 'react-ocean-forms';

/**
 * Props for components that use the info
 * feature of the FieldLine
 */
export interface IInfoProps {
  /**
   * Info text
   */
  info?: string;
}

/**
 * Base props for the FieldLine component
 */
export interface IBaseFieldLineProps extends IInfoProps {
  /**
   * Label of the input element
   */
  label: string;
  /**
   * Label size for the bootstrap label column
   */
  labelSize?: string;
  /**
   * Input size for the bootstrap input column
   */
  inputSize?: string;
  /**
   * CssClass for the label column
   */
  labelClass?: string;
  /**
   * CssClass for the field line container
   */
  containerClass?: string;

  /**
   * Prefix component
   */
  prefix?: React.ReactNode;
  /**
   * Suffix component
   */
  suffix?: React.ReactNode;
}

/**
 * Props for the FieldLine component
 */
export interface IFieldLineProps extends IUseFieldResult<unknown>, IBaseFieldLineProps { }
