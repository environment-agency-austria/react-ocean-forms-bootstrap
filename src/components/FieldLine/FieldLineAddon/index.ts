export * from './FieldLineAddon';
export * from './FieldLineAddon.types';
