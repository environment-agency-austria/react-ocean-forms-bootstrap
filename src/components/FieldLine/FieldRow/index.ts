export * from './FieldRow';
export * from './FieldRow.types';
