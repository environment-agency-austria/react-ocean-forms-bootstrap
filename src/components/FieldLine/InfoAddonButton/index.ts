export * from './InfoAddonButton';
export * from './InfoAddonButton.types';
