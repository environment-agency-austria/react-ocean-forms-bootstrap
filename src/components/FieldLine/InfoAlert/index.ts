export * from './InfoAlert';
export * from './InfoAlert.types';
