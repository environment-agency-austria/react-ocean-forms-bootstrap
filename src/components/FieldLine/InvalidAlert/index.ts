export * from './InvalidAlert';
export * from './InvalidAlert.types';
