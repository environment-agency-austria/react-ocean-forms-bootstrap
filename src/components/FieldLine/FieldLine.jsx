/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import PropTypes from 'prop-types';
import { Label, Col, InputGroup } from 'reactstrap';
import { FormText } from 'react-ocean-forms';
import { fieldMetaShape, fieldShape } from 'react-ocean-forms-legacy';

import { FieldRow } from './FieldRow';
import { RequiredMarker } from './RequiredMarker';
import { InvalidAlert } from './InvalidAlert';
import { ValidatingSpinner } from './ValidatingSpinner';
import { FieldLineAddon } from './FieldLineAddon';
import { InfoAddonButton } from './InfoAddonButton';
import { InfoAlert } from './InfoAlert';
import FieldError from './FieldError';

/**
 * Component for displaying bootstrap
 * form groups with any children
 */
class FieldLine extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      infoVisible: false,
    };

    this.toggleInfo = this.toggleInfo.bind(this);
  }

  /**
   * Toggles the visibility of the info alert
   */
  toggleInfo() {
    this.setState(prevState => ({
      infoVisible: !prevState.infoVisible,
    }));
  }

  render() {
    const {
      field,
      meta,
      label,
      children,
      validators,
      prefix,
      suffix,
      info,
      labelSize,
      inputSize,
      labelClass,
    } = this.props;

    const { infoVisible } = this.state;

    return (
      <FieldRow meta={meta}>
        <Label sm={labelSize} for={field.id} className={labelClass}>
          <InvalidAlert valid={meta.valid} />
          <ValidatingSpinner isValidating={meta.isValidating} />
          <FormText text={label} />
          <RequiredMarker meta={meta} validators={validators} />
        </Label>
        <Col sm={inputSize}>
          <InputGroup>
            <FieldLineAddon plaintext={meta.plaintext} type="prepend" content={prefix} />
            {children}
            <FieldLineAddon plaintext={meta.plaintext} type="append" content={suffix} />
            <InfoAddonButton info={info} plaintext={meta.plaintext} onClick={this.toggleInfo} />
            <FieldError
              id={`${field.id}_errors`}
              invalid={!meta.valid}
              error={meta.error}
            />
          </InputGroup>
          <InfoAlert
            visible={infoVisible}
            info={info}
            plaintext={meta.plaintext}
            onClose={this.toggleInfo}
          />
        </Col>
      </FieldRow>
    );
  }
}

FieldLine.displayName = 'FieldLine';

FieldLine.defaultProps = {
  info: undefined,
  validators: undefined,
  prefix: undefined,
  suffix: undefined,
  labelSize: 3,
  inputSize: 9,
  labelClass: 'text-right',
};

FieldLine.propTypes = {
  label: PropTypes.string.isRequired,
  info: PropTypes.string,
  meta: fieldMetaShape.isRequired,
  field: fieldShape.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  validators: PropTypes.arrayOf(PropTypes.func),
  prefix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  suffix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  labelSize: PropTypes.string,
  inputSize: PropTypes.string,
  labelClass: PropTypes.string,
};

export default FieldLine;
