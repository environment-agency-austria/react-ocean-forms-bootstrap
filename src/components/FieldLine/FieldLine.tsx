/**
 * Copyright (c) 2018-present, Umweltbundesamt GmbH
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as React from 'react';

import { FormText, PropsOf } from 'react-ocean-forms';
import { Col, InputGroup, Label } from 'reactstrap';

import { FieldError } from './FieldError';
import { FieldLineAddon } from './FieldLineAddon';
import { FieldRow } from './FieldRow';
import { InfoAddonButton } from './InfoAddonButton';
import { InfoAlert } from './InfoAlert';
import { InvalidAlert } from './InvalidAlert';
import { RequiredMarker } from './RequiredMarker';
import { ValidatingSpinner } from './ValidatingSpinner';

import { IFieldLineProps } from './FieldLine.types';

interface IFieldLineState {
  infoVisible: boolean;
}

/**
 * Component for displaying bootstrap
 * form groups with any children
 */
export class FieldLine extends React.Component<IFieldLineProps, IFieldLineState> {
  public static displayName: string = 'FieldLine';

  public static defaultProps = {
    labelSize: '3',
    inputSize: '9',
    labelClass: 'text-right',
  };

  constructor(props: IFieldLineProps) {
    super(props);

    this.state = {
      infoVisible: false,
    };
  }

  /**
   * Toggles the visibility of the info alert
   */
  private toggleInfo = (): void => {
    this.setState(prevState => ({
      infoVisible: !prevState.infoVisible,
    }));
  }

  public render(): JSX.Element {
    const {
      field,
      meta,
      label,
      children,
      prefix,
      suffix,
      info,
      labelSize,
      inputSize,
      labelClass,
    } = this.props;

    const { infoVisible } = this.state;

    return (
      <FieldRow meta={meta}>
        <Label sm={labelSize} for={field.id} className={labelClass}>
          <InvalidAlert valid={meta.valid} />
          <ValidatingSpinner isValidating={meta.isValidating} />
          <FormText text={label} />
          <RequiredMarker meta={meta} />
        </Label>
        <Col sm={inputSize}>
          <InputGroup>
            <FieldLineAddon plaintext={meta.plaintext} type="prepend" content={prefix} />
            {children}
            <FieldLineAddon plaintext={meta.plaintext} type="append" content={suffix} />
            <InfoAddonButton info={info} plaintext={meta.plaintext} onClick={this.toggleInfo} />
            <FieldError
              id={`${field.id}_errors`}
              invalid={!meta.valid}
              error={meta.error}
            />
          </InputGroup>
          <InfoAlert
            visible={infoVisible}
            info={info}
            plaintext={meta.plaintext}
            onClose={this.toggleInfo}
          />
        </Col>
      </FieldRow>
    );
  }
}

export type DefaultizedFieldLineProps = JSX.LibraryManagedAttributes<typeof FieldLine, PropsOf<typeof FieldLine>>;
