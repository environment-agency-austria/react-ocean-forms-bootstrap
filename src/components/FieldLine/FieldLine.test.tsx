import React from 'react';

import { render, fireEvent, waitForElementToBeRemoved } from '@testing-library/react';
import { Form } from 'react-ocean-forms';

import { FieldLine } from './FieldLine';
import { IFieldLineProps } from './FieldLine.types';

describe('<FieldLine />', () => {
  function generateMockProps(): IFieldLineProps {
    return {
      fieldProps: {
        id: 'mock',
        name: 'mock',
        value: 'mock',
        disabled: false,
        onChange: jest.fn(),
        onBlur: jest.fn(),
      },
      metaProps: {
        touched: false,
        plaintext: false,
        stringFormatter: jest.fn().mockImplementation((val) => val),
        isRequired: false,
        isValidating: false,
        valid: true,
        error: null,
      },
      label: 'mock',
    };
  }

  it('should render correctly', () => {
    const props = generateMockProps();
    const { asFragment, getByText } = render(
      <Form>
        <FieldLine {...props} />
      </Form>
    );

    expect(asFragment()).toMatchSnapshot();
    expect(getByText('mock')).toBeVisible();
  });

  it('should correctly set custom prop values', () => {
    const props = generateMockProps();
    const { asFragment } = render(
      <Form>
        <FieldLine {...props} labelSize="2" inputSize="10" labelClass="custom" containerClass="custom container" />
      </Form>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  describe('prefix and suffix', () => {
    it('should display a prefix if provided', () => {
      const props = generateMockProps();
      const { getByText } = render(
        <Form>
          <FieldLine {...props} prefix="Prefix" />
        </Form>
      );

      expect(getByText('Prefix')).toBeVisible();
    });

    it('should display a suffix if provided', () => {
      const props = generateMockProps();
      const { getByText } = render(
        <Form>
          <FieldLine {...props} suffix={<div>Suffix</div>} />
        </Form>
      );

      expect(getByText('Suffix')).toBeVisible();
    });
  });

  describe('info handling', () => {
    it('should not display an info button without info', () => {
      const props = generateMockProps();
      const { queryByTitle } = render(
        <Form>
          <FieldLine {...props} />
        </Form>
      );

      expect(queryByTitle('ojs_show_information')).toBeNull();
    });

    it('should display an info button', () => {
      const props = generateMockProps();
      const { getByTitle } = render(
        <Form>
          <FieldLine {...props} info="mock info" />
        </Form>
      );

      expect(getByTitle('ojs_show_information')).toBeVisible();
    });

    it('should togle info visibility when info button is clicked', async () => {
      const props = generateMockProps();
      const { getByText, queryByText, getByTitle } = render(
        <Form>
          <FieldLine {...props} info="mock info" />
        </Form>
      );

      expect(queryByText('mock info')).toBeNull();

      fireEvent.click(getByTitle('ojs_show_information'));
      expect(getByText('mock info')).toBeVisible();

      fireEvent.click(getByTitle('ojs_show_information'));
      await waitForElementToBeRemoved(() => getByText('mock info'));
      expect(queryByText('mock info')).toBeNull();
    });
  });

  describe('error handling', () => {
    it('should not have any error roles in default mode', () => {
      const props = generateMockProps();
      const { getByRole } = render(
        <Form>
          <FieldLine {...props} />
        </Form>
      );

      expect(getByRole('form').firstChild).not.toHaveClass('is-invalid');
    });

    it('should correctly display a single error', () => {
      const props = generateMockProps();
      props.metaProps.valid = false;
      props.metaProps.error = { message_id: 'mock error', params: { } };

      const { getByText, getByRole } = render(
        <Form>
          <FieldLine {...props} />
        </Form>
      );

      expect(getByText('mock error')).toBeVisible();
      expect(getByRole('form').firstChild).toHaveClass('is-invalid');
    });

    it('should correctly display multiple errors', () => {
      const props = generateMockProps();
      props.metaProps.valid = false;
      props.metaProps.error = [
        { message_id: 'mock error', params: { } },
        { message_id: 'mock second error', params: { } }
      ];

      const { getByText } = render(
        <Form>
          <FieldLine {...props} />
        </Form>
      );

      expect(getByText('mock error')).toBeVisible();
      expect(getByText('mock second error')).toBeVisible();
    });
  });

  describe('required fields handling', () => {
    it('should not display a required marker on not required fields', () => {
      const props = generateMockProps();
      const { queryByTitle } = render(
        <Form>
          <FieldLine {...props} />
        </Form>
      );

      expect(queryByTitle('Required')).toBeNull();
    });

    it('should display a required marker on required fields', () => {
      const props = generateMockProps();
      props.metaProps.isRequired = true;

      const { getByTitle } = render(
        <Form>
          <FieldLine {...props} />
        </Form>
      );

      expect(getByTitle('Required')).toBeVisible();
    });

    it('should not display a required marker on required fields in plaintext mode', () => {
      const props = generateMockProps();
      props.metaProps.isRequired = true;
      props.metaProps.plaintext = true;

      const { queryByTitle } = render(
        <Form>
          <FieldLine {...props} />
        </Form>
      );

      expect(queryByTitle('Required')).toBeNull();
    });
  });

  describe('validation handling', () => {
    it('should not display a validating spinner on not validating fields', () => {
      const props = generateMockProps();
      const { queryByTitle } = render(
        <Form>
          <FieldLine {...props} />
        </Form>
      );

      expect(queryByTitle('ojs_field_validating')).toBeNull();
    });

    it('should display a validating spinner on validating fields', () => {
      const props = generateMockProps();
      props.metaProps.isValidating = true;

      const { getByTitle } = render(
        <Form>
          <FieldLine {...props} />
        </Form>
      );

      expect(getByTitle('ojs_field_validating')).toBeInTheDocument();
    });
  });

  describe('touched handling', () => {
    it('should not have is-touched class by default', () => {
      const props = generateMockProps();
      const { getByRole } = render(
        <Form>
          <FieldLine {...props} />
        </Form>
      );

      expect(getByRole('form').firstChild).not.toHaveClass('is-touched');
    });

    it('should have the is-touched class on touched fields', () => {
      const props = generateMockProps();
      props.metaProps.touched = true;

      const { getByRole } = render(
        <Form>
          <FieldLine {...props} />
        </Form>
      );

      expect(getByRole('form').firstChild).toHaveClass('is-touched');
    });
  });
});
