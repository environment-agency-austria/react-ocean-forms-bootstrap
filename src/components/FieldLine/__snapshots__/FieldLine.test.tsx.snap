// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldLine /> should render correctly 1`] = `
<FieldRow
  meta={
    Object {
      "error": null,
      "isRequired": false,
      "isValidating": false,
      "plaintext": false,
      "stringFormatter": [MockFunction],
      "touched": false,
      "valid": true,
    }
  }
>
  <Label
    className="text-right"
    for="mock-field"
    sm="3"
    tag="label"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <InvalidAlert
      valid={true}
    />
    <ValidatingSpinner
      isValidating={false}
    />
    <FormText
      text="label"
    />
    <RequiredMarker
      meta={
        Object {
          "error": null,
          "isRequired": false,
          "isValidating": false,
          "plaintext": false,
          "stringFormatter": [MockFunction],
          "touched": false,
          "valid": true,
        }
      }
    />
  </Label>
  <Col
    sm="9"
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
  >
    <InputGroup
      tag="div"
    >
      <FieldLineAddon
        plaintext={false}
        type="prepend"
      />
      <div
        id="mock-child"
      />
      <FieldLineAddon
        plaintext={false}
        type="append"
      />
      <InfoAddonButton
        onClick={[Function]}
        plaintext={false}
      />
      <FieldError
        error={null}
        id="mock-field_errors"
        invalid={false}
      />
    </InputGroup>
    <InfoAlert
      onClose={[Function]}
      plaintext={false}
      visible={false}
    />
  </Col>
</FieldRow>
`;
